experimental/oak_async/src/executor.rs

```diff
@@ -0,0 +1,176 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//! Async executor for Oak Nodes.
+
+use core::{cell::RefCell, future::Future, task::Waker};
+use futures::{executor::LocalPool, task::LocalSpawnExt};
+use log::{debug, trace};
+use oak::{ChannelReadStatus, Handle, OakStatus, ReadHandle};
+use std::collections::HashMap;
+
+// thread local so `cargo test` can run tests in parallel.
+// Oak nodes are single-threaded.
+std::thread_local! {
+    static EXECUTOR: RefCell<Executor> = RefCell::new(Executor::default());
+}
+
+type ReaderId = usize;
+
+struct WaitingReader {
+    handle: Handle,
+    waker: Waker,
+}
+
+/// Global executor context.
+///
+/// `with_executor` provides a way to get a handle to the global executor instance.
+#[derive(Default)]
+pub struct Executor {
+    /// Used to assign a unique id to each reader. It is incremented by every newly created reader
+    reader_id_counter: ReaderId,
+    /// Set of readers waiting for data. All handles in this map are polled in `wait_on_channels`.
+    // Why not HashMap<Handle, Vec<Waker>>? When a future asks to be removed from the waiting list
+    // (i.e. when it is dropped), we need to be able to remove their Waker from the map.
+    waiting_readers: HashMap<ReaderId, WaitingReader>,
+}
+
+impl Executor {
+    pub fn add_waiting_reader(&mut self, reader_id: usize, handle: ReadHandle, waker: &Waker) {
+        trace!(
+            "Add waiting reader {} waiting on handle {}",
+            reader_id,
+            handle.handle
+        );
+        let _ = self.waiting_readers.insert(
```

Comment:
Why the `let _`? Is `insert` marked as `must_use`?

Reply:
I thought the `Option` would be `must_use`, but it's not. Thanks!